crates/astria-sequencer/src/accounts/query.rs

```diff
@@ -1,84 +1,180 @@
+        0 => storage.latest_snapshot(),
+        other => {
+            let version = storage
+                .latest_snapshot()
+                .get_storage_version_by_height(other)
+                .await
+                .context("failed to get storage version from height")?;
+            storage
+                .snapshot(version)
+                .context("failed to get storage at version")?
+        }
+    };
+    let height: Height = snapshot
+        .get_block_height()
+        .await
+        .context("failed to get block height from snapshot")?
+        .try_into()
+        .context("internal u64 block height does not fit into tendermint i64 `Height`")?;
+    Ok((snapshot, height))
 }
 
-#[derive(Default)]
-pub(crate) struct QueryHandler {}
-
-impl QueryHandler {
-    pub(crate) fn new() -> Self {
-        Self {}
-    }
-
-    #[instrument(skip(self, state))]
-    pub(crate) async fn handle<S: StateReadExt>(
-        &self,
-        state: S,
-        query: Request,
-    ) -> Result<Response> {
-        match query {
-            Request::BalanceQuery(address) => {
-                let balance = state
-                    .get_account_balance(&address)
-                    .await
-                    .context("failed getting account balance")?;
-                Ok(Response::BalanceResponse(balance))
-            }
-            Request::NonceQuery(address) => {
-                let nonce = state
-                    .get_account_nonce(&address)
-                    .await
-                    .context("failed getting account nonce")?;
-                Ok(Response::NonceResponse(nonce))
-            }
+async fn preprocess_request(
+    storage: &Storage,
+    request: &request::Query,
+    params: &[(String, String)],
+) -> anyhow::Result<(Address, Snapshot, Height), response::Query> {
+    let Some(address) = params
+        .iter()
+        .find_map(|(k, v)| (k == "account").then_some(v))
+    else {
+        return Err(response::Query {
+            code: AbciCode::INVALID_PARAMETER.into(),
+            info: format!("{}", AbciCode::INVALID_PARAMETER),
+            log: "path did not contain path parameter".into(),
+            ..response::Query::default()
+        });
+    };
+    let address = hex::decode(address)
+        .map_err(|err| response::Query {
+            code: AbciCode::INVALID_PARAMETER.into(),
+            info: format!("{}", AbciCode::INVALID_PARAMETER),
+            log: format!(
+                "account public key could not be constructed from provided paratemer: {err:?}"
+            ),
+            ..response::Query::default()
+        })
+        .and_then(|addr| {
+            Address::try_from_slice(&addr).map_err(|err| response::Query {
+                code: AbciCode::INVALID_PARAMETER.into(),
+                info: format!("{}", AbciCode::INVALID_PARAMETER),
+                log: format!(
+                    "account public key could not be constructed from provided paratemer: {err:?}"
```

Comment:
```suggestion
                    "address could not be constructed from provided parameter: {err:?}"
```

Reply:
used the suggested change and streamlined the response construction.